### ISSUE TITLE
Proposal for fixing issues related to coupling between model checkpoints and `.git_theta/`

### ISSUE THREAD

**Issue by nkandpa2** [r-three/git-theta#111]
# Problems
The current design couples the state of a tracked checkpoint on disk with the contents of `.git_theta/` and tracks the contents of most of `.git_theta` using git LFS. I feel like lately we've been running into issues that arise from this coupling of checkpoints and `.git_theta`. Here are a few of the issues that we've hit and general inefficiencies of doing things this way:

1. When we smudge a checkpoint, we rely on `.git_theta/` having the correct contents. This happens most of the time since git populates the working tree in lexical order and model checkpoints usually have names that come after `.git_theta` lexically. However, this is an undocumented feature that should not be relied upon.
2. When we want to diff or merge two checkpoints, their representations in `.git_theta/` are useful since they give information about what updates were applied between the two checkpoints. However, at the time that the diff and merge tools run, only one of the versions (the currently checked out version) can be in `.git_theta/`. 
3. When we stage a model that is $n$ bytes on disk, the effective size of the repo is about $3n$ bytes which seems unnecessary (one copy in the working tree, one copy in `.git_theta/`, and one copy in `.git/lfs` since the contents of `.git_theta/` are tracked by LFS)
4. Probably some issues in the future

The core of some of many of these issues is that git is not meant to be used in this way where each version of a file is tied to a corresponding version of other files, since git treats each file independently for all of its operations (checkout, diff, merge, etc.). Here is a proposal to fix these issues.

# Implementation
#### Upon `git add <checkpoint>` the clean filter should:
1. Load the checkpoint into memory
2. Write each parameter group to a temporary file
3. Pass each temporary file to `git lfs clean`. This command will copy the contents of the file to `.git/lfs/objects` and output the contents of an LFS pointer file to stdout.
4. Save the output of `git lfs clean` (i.e., the contents of the pointer files). This output contains the git LFS spec version, the file's LFS oid (object id), and the file size.
5. Create a metadata file of the following form:
```json
{
    "param_group1": {
        "oid": "object id 1",
        "size": 1000
    }
    "param_group2": {
        "oid": "object id 2",
        "size": 500
    }
}
```
Basically each parameter group name maps to a dictionary containing the information that git LFS would normally put in that parameter group's pointer file.

#### Upon `git push` the pre-push hook should:
1. Iterate over all files being committed
2. Find the git-theta tracked files (i.e., see if they match the git-theta metadata file format)
3. For each of the git-theta tracked files, run the command `git lfs push --object-id origin <list of object-ids>` where the list of object-ids is the object-ids in the metadata file

#### Upon `git checkout` the smudge filter should:
1. Read the metadata file
2. For each parameter group create a temporary git LFS pointer file with the object id and size of that parameter group
3. Pass the pointer file to `git lfs smudge`. This command will check `.git/lfs/objects` to see if the object is already on disk locally and if it isn't will download it from the LFS server. The contents of the object will be written to stdout.
4. Save the output of `git lfs smudge` to a temporary file
6. Load the temporary file to get the parameter values
7. Reconstitute the model checkpoint 

# Intuition
Earlier on we'd discussed forking Git LFS to add a feature where a single pointer file can point to multiple LFS objects. This proposal is essentially doing that by translating between a model's metadata file and a set of temporary LFS pointer files that get passed to Git LFS. 

# Advantages of this Approach
1. We no longer have the coupling between checkpoints and `.git_theta/`
2. Users can run `git add <checkpoint>` instead of `git theta add <checkpoint>`
3. Effective size of the repo is $2n$ bytes after staging a checkpoint that is $n$ bytes (one copy in working tree, one copy in .git/lfs`

**Comment by craffel** [comment]
Thanks for writing this - would the proposed approach fix this issue?
>  However, at the time that the diff and merge tools run, only one of the versions (the currently checked out version) can be in .git_theta/.

**Comment by nkandpa2** [comment]
Yeah, say you want to diff two commits from the past. To get either of the commits or any intermediate commits between them we would just get the commit hash with `git log`, get the contents of the metadata file at that commit with `git show`, and then smudge that file with `git-theta-filter smudge`. The reason this is currently difficult is because to smudge a metadata file at a certain commit, we also need the state in the working tree under `.git_theta/` to match the commit of the metadata file we're smudging.

**Comment by craffel** [comment]
Sounds cool to me! Happy to discuss more but I'm also in support in you giving it a stab.

**Comment by nkandpa2** [comment]
Closed by #114
